### PR TITLE
Allow legacy index settings on legacy indices

### LINKED ISSUE #84992: Investigate severity of upgrading to 8.0.1 despite terminal deprecation of old features
### Elasticsearch Version

upgrade from 7.x -> 8.0.0

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux

### Problem Description

Recently we encountered a user issue report where upgrades to 8.0.1 from 7.17.1 caused an unhealthy cluster because the option `index.indexing.slowlog.level` wasn't removed prior to upgrading.

We'll use this issue to understand how this happened and ways we can prevent the upgrade before irreversible changes are done to the Elasticsearch metadata or data.

### Steps to Reproduce

- Get Elasticsearch 7.9 cluster with the `index.indexing.slowlog.level` option set. 
- Upgrade to 7.17.1 ignore the deprecation warning
- Upgrade to 8.0.1

Expected behaviour: the 8.0.1 upgrade should fail to start but not upgrade any metadata or data
Current behaviour: the upgrade to 8.0.1 proceeds irreversibly modifying the metadata before options are validated and start is unsuccessful.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by grcevski**
Our backward compatibility on indices states that we support indices with one version less than the current version. Those legacy indices might have legacy index settings, which we typically archive on cluster restart. However, this archiving happens only on full cluster restart, not on rolling restarts, which effectively leaves the indices unusable in 8.x.

This PR adds code that ignores unknown settings on indices if the index created version is less than the current version.

Closes https://github.com/elastic/elasticsearch/issues/84992

### CODE REVIEW

**Review by DaveCTurner — DISMISSED** [review]
I think we shouldn't introduce a broad-brush lenience like this. We should instead reinstate the settings that we removed, along with a setting validator that prevents them from being set on 8.x indices.

By leaving these settings unregistered I think we prevent folks from adjusting or removing them on older indices, but these things should remain possible even in 8.x. I also think they'd still be archived in a full-cluster restart but we don't want that either.

(clarifying edit:) In the case of the slowlog settings I think it's ok if setting them has no effect in 8.x. But they still should remain registered.

**Comment by grcevski** [comment]
> We should instead reinstate the settings that we removed, along with a setting validator that prevents them from being set on 8.x indices.

Good idea! I'll take that approach instead, it's similar to what we had done for camel case date field support.

**Comment by grcevski** [comment]
@DaveCTurner, I thought more about this and I think we should be archiving these settings before validation. If we silently accept them as valid settings for 7.x indices, we then run into a problem where the usage will be confusing to our end users.

For example, the slow logs setting did something in 7.x, even though we warned them of deprecation, but now in 8.x we will allow it to exist and be changed as a setting, but it will do nothing. So it's sort of valid setting you can set, but it does nothing anymore. It still breaks BWC, I'd rather we told them somehow their setting no longer works.

I think a better solution would be to archive the legacy settings when we validate them. I can bring back the list of these legacy settings and only archive those, but we archive all unknown settings already on restart, I think we might as well archive all of them.

**Comment by DaveCTurner** [comment]
I would very much prefer not to archive any index settings. We almost never do a full-cluster-restart upgrade any more, and during a rolling upgrade it is not at all obvious when to do the archiving, nor how we should behave in the mixed-version cluster before the archival happens.

It is unfortunate that in this case we removed the associated slowlog functionality prematurely, but ISTR we had good reasons for doing so. I think that won't cause too many problems however: the settings were properly deprecated in 7.x so the user should have already seen guidance about their use, and as long as the settings are still marked as deprecated in 8.x the user will continue to see warnings about not using them. We could even perhaps customise those warnings to indicate that they're already no-ops, although I wouldn't put much effort into that until we start to see evidence that it's needed.

Are there other settings to be reinstated or is it just about these slowlog ones? If there are others, we'll need to think about what to do with them on a case-by-case basis.

File: server/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 1]
```diff
@@ -139,7 +139,12 @@ public enum Property {
         /**
          * Indicates an index-level setting that is privately managed. Such a setting can not even be set on index creation.
          */
-        PrivateIndex
+        PrivateIndex,
+
+        /**
+         * Indicates that this index-level setting was deprecated and removed in the current major
+         */
+        DeprecatedAndRemovedInCurrentMajor
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/settings/Setting.java:147]: Likewise here, let's mention the specific versions in the symbol name and trigger a compile error in v9. Also this property only makes sense for index settings. How about something like this:

```suggestion
        /**
         * Indicates that this index-level setting was deprecated in {@link Version#V_7_17_0) and removed in {@link Version#V_8_0_0}.
         */
        IndexSettingDeprecatedInV7AndRemovedInV8
```

Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/settings/Setting.java:147]: We should also reject non-index settings that use this property - see `checkPropertyRequiresIndexScope`. Also I think we can make this property mutually exclusive with `Deprecated` and `DeprecatedWarning`.

Also (maybe in a followup) can we forbid `Deprecated` on `IndexScope` settings and require all index settings to use this framework instead?

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 2]
```diff
@@ -550,6 +551,35 @@ public Iterator<Setting<?>> settings() {
...
+            public void validate(Integer value, Map<Setting<?>, Object> settings) {
+                validateVersionDependentDeprecatedSetting(MAX_ADJACENCY_MATRIX_FILTERS_SETTING.getKey(), settings);
+            }
+
+            @Override
+            public Iterator<Setting<?>> settings() {
+                final List<Setting<?>> settings = List.of(IndexMetadata.SETTING_INDEX_VERSION_CREATED);
+                return settings.iterator();
+            }
+        },
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:577]: Hmm I think we should do this validation automatically within the settings framework if a setting is marked as `Property.DeprecatedAndRemovedInCurrentMajor`. If we require implementors to remember to include this validator too then someone will forget and that'd be a bug.

**Review by DaveCTurner — COMMENTED** [review]
I left some thoughts mostly on code-habitability. I like this approach much more tho.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/UpgradeWithOldIndexSettingsIT.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.rest.action.search.RestSearchAction.TOTAL_HITS_AS_INT_PARAM;
+import static org.hamcrest.Matchers.is;
+
+public class UpgradeWithOldIndexSettingsIT extends AbstractRollingTestCase {
```
Comment (pgomulka) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/UpgradeWithOldIndexSettingsIT.java:26]: out of curiosity.
Any particular reason not to use yml test for this?

File: server/src/main/java/org/elasticsearch/common/settings/Setting.java (hunk 1 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/settings/Setting.java:147]: I wonder if we have to mention Deprecation in the property name.
howe about `Property.ToBeRemovedInV8IndexSetting`? or `Property.IndexSettingToBeRemovedInV8`?

**Review by pgomulka — APPROVED** [review]
I like the approach and left minor questions.

File: server/src/main/java/org/elasticsearch/common/settings/Setting.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/settings/Setting.java:147]: I could live with that too, maybe even just `IndexSettingRemovedInV8`, although I wonder if it's helpful to be explicit about the property's effects in its name. We must mention `Version#V_7_17_0` in the javadoc to ensure this property is removed in v9.

Comment (grcevski) [reply server/src/main/java/org/elasticsearch/common/settings/Setting.java]: Good idea. About removing all Deprecated, I guess we'll need one more property, deprecated and removed in Next major, if we have deprecated any in 8.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 2 above)
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:577]: Fair enough, I can add the validation at lower level.

File: server/src/main/java/org/elasticsearch/common/settings/Setting.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/settings/Setting.java]: By my reckoning there's currently exactly one such index setting that's deprecated in v8 and to be removed in v9: `index.merge.policy.max_merge_at_once_explicit`.

I'll leave it up to you whether to give this setting the same treatment now or later.

**Comment by grcevski** [comment]
@elasticsearchmachine run elasticsearch-ci/docs

**Review by DaveCTurner — COMMENTED** [review]
Looks good to me; I left one optional suggestion

File: server/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 4]
```diff
@@ -184,9 +190,14 @@ private Setting(
             if (propertiesAsSet.contains(Property.Deprecated) && propertiesAsSet.contains(Property.DeprecatedWarning)) {
                 throw new IllegalArgumentException("setting [" + key + "] cannot be deprecated at both critical and warning levels");
             }
+            if ((propertiesAsSet.contains(Property.Deprecated) || propertiesAsSet.contains(Property.DeprecatedWarning))
+                && propertiesAsSet.contains(Property.IndexSettingDeprecatedInV7AndRemovedInV8)) {
+                throw new IllegalArgumentException("setting [" + key + "] cannot be just deprecated and deprecated and removed in V7");
+            }
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/settings/Setting.java:196]: Hmm suggestion got lost it seems. Here it is again:

```suggestion
            if (propertiesAsSet.stream().filter(DEPRECATED_PROPERTIES::contains).count() > 1) {
                throw new IllegalArgumentException("setting [" + key + "] must be at most one of [" + DEPRECATED_PROPERTIES + "]");
            }
```

Comment (grcevski) [reply server/src/main/java/org/elasticsearch/common/settings/Setting.java:196]: That's a really nice code simplification. I'll add it in.

**Pull request merged**